test/integration/master/audit_test.go

```diff
@@ -0,0 +1,354 @@
+					"watch",
+					200,
+					auditTestUser,
+					"configmaps",
+					namespace,
+					false,
+					false,
+					"allow",
+				}, {
+					auditinternal.LevelRequestResponse,
+					auditinternal.StageResponseComplete,
+					fmt.Sprintf("/api/v1/namespaces/%s/configmaps?timeoutSeconds=%d&watch=true", namespace, watchTestTimeout),
+					"watch",
+					200,
+					auditTestUser,
+					"configmaps",
+					namespace,
+					false,
+					false,
+					"allow",
+				}, {
+					auditinternal.LevelRequestResponse,
+					auditinternal.StageResponseComplete,
+					fmt.Sprintf("/api/v1/namespaces/%s/configmaps/audit-configmap", namespace),
+					"update",
+					200,
+					auditTestUser,
+					"configmaps",
+					namespace,
+					true,
+					true,
+					"allow",
+				}, {
+					auditinternal.LevelRequestResponse,
+					auditinternal.StageResponseComplete,
+					fmt.Sprintf("/api/v1/namespaces/%s/configmaps/audit-configmap", namespace),
+					"patch",
+					200,
+					auditTestUser,
+					"configmaps",
+					namespace,
+					true,
+					true,
+					"allow",
+				}, {
+					auditinternal.LevelRequestResponse,
+					auditinternal.StageResponseComplete,
+					fmt.Sprintf("/api/v1/namespaces/%s/configmaps/audit-configmap", namespace),
+					"delete",
+					200,
+					auditTestUser,
+					"configmaps",
+					namespace,
+					true,
+					true,
+					"allow",
+				},
+			},
+		},
+	}
+
+	expectedEvents := []auditEvent{}
+	for _, t := range testCases {
+		t.action()
+		expectedEvents = append(expectedEvents, t.events...)
+	}
+	ok, err := checkAuditLines(t, logFile.Name(), expectedEvents, version)
+	if err != nil {
+		t.Fatalf("Unexpected error: %v", err)
+	}
+	if !ok {
+		t.Fatalf("Failed to match all expected events")
+	}
+}
+
+type auditEvent struct {
+	level             auditinternal.Level
+	stage             auditinternal.Stage
+	requestURI        string
+	verb              string
+	code              int32
+	user              string
+	resource          string
+	namespace         string
+	requestObject     bool
+	responseObject    bool
+	authorizeDecision string
+}
+
+func expectNoError(t *testing.T, err error, msg string) {
+	if err != nil {
+		t.Fatalf("%s: %v", msg, err)
+	}
+}
+
+func checkAuditLines(t *testing.T, logFile string, expected []auditEvent, version string) (bool, error) {
+	expectations := map[auditEvent]bool{}
+	for _, event := range expected {
+		expectations[event] = false
+	}
+
+	// Fetch the log stream.
+	stream, err := os.Open(logFile)
+	if err != nil {
+		return false, err
+	}
+	defer stream.Close()
+
+	scanner := bufio.NewScanner(stream)
+	for scanner.Scan() {
+		line := scanner.Text()
+		event, err := parseAuditLine(line, version)
+		if err != nil {
+			return false, err
+		}
+
+		// If the event was expected, mark it as found.
+		if _, found := expectations[event]; found {
+			expectations[event] = true
+		}
+	}
+	if err := scanner.Err(); err != nil {
+		return false, err
+	}
+
+	noneMissing := true
```

Comment:
s/noneMissing/allFound/? Double negation is harder to read.

Reply:
done.